book/src/dev/rfcs/0005-treestate.md

```diff
@@ -0,0 +1,193 @@
+previous `JoinSplit`s in that transaction finish validating before doing the
+anchor consensus check again, but against the output treestate roots of earlier
+`JoinSplit`s.
+
+Both Sprout and Sapling `NoteCommitment` trees must be computed for the whole
+block to validate, but Sprout, we need to compute interstitial treestates in
+between `JoinSplit`s in order to do the final consensus check for each/all
+`JoinSplit`s, not just for the whole block, like in Sapling.
+
+For Sapling, at the block layer, we can iterate over all the transactions in
+order and if they have `Spend`s and/or `Output`s, we update our Nullifer set for
+the block as nullifiers are revealed in `Spend` descriptions, and update our note
+commitment tree as `NoteCommitment`s are revealed in `Output` descriptions, adding
+them as leaves in positions according to their order as they appear transaction
+to transction, output to output, in the block. This can be done independent of
+the transaction validations. When the Sapling transactions are all validated,
+the `NoteCommitmentTree` root should be computed: this is the anchor for this
+block. For Sapling and Blossom blocks, we need to check that this root matches
+the `RootHash` bytes in this block's header, as the `FinalSaplingRoot`. Once all
+other consensus and validation checks are done, this will be saved down to our
+finalized state to our `sapling_anchors` set, making it available for lookup by
+other Sapling descriptions in future transactions.
+
+For Sprout, we must compute/update interstitial `NoteCommitmentTree`s between
+`JoinSplit`s that may reference an earlier one's root as its anchor. If we do
+this at the transaction layer, we can iterate through all the `JoinSplit`s and
+compute the Sprout `NoteCommitmentTree` and nullifier set similar to how we do
+the Sapling ones as described above, but at each state change (ie,
+per-`JoinSplit`) we note the root and cache it for lookup later. As the
+`JoinSplit`s are validated without context, we check for its specified anchor
+amongst the interstitial roots we've already calculated (according to the spec,
+these interstitial roots don't have to be finalized or the result of an
+independently validated `JoinSplit`, they just must refer to any prior `JoinSplit`
+root in the same transaction). So we only have to wait for our previous root to
+be computed via any of our candidates, which in the worst case is waiting for
+all of them to be computed for the last `JoinSplit`. If our `JoinSplit`s defined
+root pops out, that `JoinSplit` passes that check.
+
+To finalize the block, the Sprout and Sapling treestates are the ones resulting
+from the last transaction in the block, and determines the Sprout and Sapling
+anchors that will be associated with this block as we commit it to our finalized
+state. The Sprout and Sapling nullifiers revealed in the block will be merged
+with the exising ones in our finalized state (ie, it should strictly grow over
+time).
+
+
```

Comment:
IMPORTANT: we need to save the incremental merkle tree / serialized nodes for:

- Sapling tip block
- ~all Sprout blocks
- chains we're tracking in memory within the reorg limit, for both

Reply:
Added https://github.com/ZcashFoundation/zebra/pull/983/commits/67c7f49972be486cbb1ce0a6d2f46319ba103ac1